From 683f1ad81823800271d6f000e3e8b2155a3e386a Mon Sep 17 00:00:00 2001
From: Valery Ushakov <valery.ushakov@bell-sw.com>
Date: Fri, 22 Aug 2025 10:52:30 +0000
Subject: [PATCH 2/2] read_cmdline: wrap in printable_string

This is the same issue as CVE-2024-58251 for netstat -p vs. control
characters in argv0.  read_cmdline() had rudimentary sanitation code
that masked C0 controls, but there are C1 controls too and gnome
terminal seems to always interpret them, and xterm interprets them in
non-utf-8 mode.  So wrap read_cmdline() in a piece of advice that
applies printable_string to the buffer and add init_unicode() calls to
ps(1), top(1) and pmap(1) that call read_cmdline().

Note that this is not _completely_ fool-proof as you may have C1 bytes
as part of valid UTF-8 sequences for characters outside C1, e.g.
\u3add0;Gotcha\u3adc is a valid UTF-8 string, but it also a valid OSC
sequence _if_ interpreted as a single-byte 8-bit string.  procps-ng
will pass that kind of sequence through as well.  But this is,
strictly speaking, a misconfiguration, and you are supposed to use
Ubik's self-winding Swiss chromium never-ending blade only as
directed, and with caution.

Signed-off-by: Valery Ushakov <valery.ushakov@bell-sw.com>
---
 libbb/procps.c | 16 ++++++++++++++--
 procps/pmap.c  |  3 +++
 procps/ps.c    |  2 ++
 procps/top.c   |  2 ++
 4 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/libbb/procps.c b/libbb/procps.c
index f56b71b21..f8aae9af0 100644
--- a/libbb/procps.c
+++ b/libbb/procps.c
@@ -566,7 +566,7 @@ procps_status_t* FAST_FUNC procps_scan(procps_status_t* sp, int flags)
 	return sp;
 }
 
-void FAST_FUNC read_cmdline(char *buf, int col, unsigned pid, const char *comm)
+static void FAST_FUNC read_cmdline2(char *buf, int col, unsigned pid, const char *comm)
 {
 	int sz;
 	char filename[sizeof("/proc/%u/cmdline") + sizeof(int)*3];
@@ -584,7 +584,8 @@ void FAST_FUNC read_cmdline(char *buf, int col, unsigned pid, const char *comm)
 		strchrnul(buf, ' ')[0] = '\0';
 		base = bb_basename(buf); /* before we replace argv0's NUL with space */
 		while (sz >= 0) {
-			if ((unsigned char)(buf[sz]) < ' ')
+			/* controls will be dealt with by printable_string */
+			if (buf[sz] == '\0')
 				buf[sz] = ' ';
 			sz--;
 		}
@@ -618,6 +619,17 @@ void FAST_FUNC read_cmdline(char *buf, int col, unsigned pid, const char *comm)
 	}
 }
 
+// defadvice (read_cmdline :after)
+void FAST_FUNC read_cmdline(char *buf, int col, unsigned pid, const char *comm)
+{
+	read_cmdline2(buf, col, pid, comm);
+	char *printable = printable_string(buf);
+	if (printable == buf)
+		return;
+
+	snprintf(buf, col, "%s", printable);
+}
+
 /* from kernel:
 	//             pid comm S ppid pgid sid tty_nr tty_pgrp flg
 	sprintf(buffer,"%d (%s) %c %d  %d   %d  %d     %d       %lu %lu \
diff --git a/procps/pmap.c b/procps/pmap.c
index 49f7688d9..3b367d34c 100644
--- a/procps/pmap.c
+++ b/procps/pmap.c
@@ -26,6 +26,7 @@
 //usage:     "\n	-q	Quiet"
 
 #include "libbb.h"
+#include "unicode.h"
 
 #if ULLONG_MAX == 0xffffffff
 # define TABS "\t"
@@ -96,6 +97,8 @@ int pmap_main(int argc UNUSED_PARAM, char **argv)
 	unsigned opts;
 	int ret;
 
+	init_unicode();
+
 	opts = getopt32(argv, "^" "xq" "\0" "-1"); /* min one arg */
 	argv += optind;
 
diff --git a/procps/ps.c b/procps/ps.c
index 5b521aebd..f184f25ac 100644
--- a/procps/ps.c
+++ b/procps/ps.c
@@ -109,6 +109,7 @@
 //usage:       " 2990 andersen andersen R ps\n"
 
 #include "libbb.h"
+#include "unicode.h"
 #include "common_bufsiz.h"
 #ifdef __linux__
 # include <sys/sysinfo.h>
@@ -576,6 +577,7 @@ int ps_main(int argc UNUSED_PARAM, char **argv)
 	G.kernel_HZ = bb_clk_tck(); /* this is sysconf(_SC_CLK_TCK) */
 # endif
 #endif
+	init_unicode();
 
 	// POSIX:
 	// -a  Write information for all processes associated with terminals
diff --git a/procps/top.c b/procps/top.c
index 09d31c673..64acbafcb 100644
--- a/procps/top.c
+++ b/procps/top.c
@@ -116,6 +116,7 @@
 //kbuild:lib-$(CONFIG_TOP) += top.o
 
 #include "libbb.h"
+#include "unicode.h"
 
 #define ESC "\033"
 
@@ -1160,6 +1161,7 @@ int top_main(int argc UNUSED_PARAM, char **argv)
 	unsigned scan_mask = TOP_MASK;
 
 	INIT_G();
+	init_unicode();
 
 	interval = 5; /* default update interval is 5 seconds */
 	iterations = 0; /* infinite */
-- 
2.34.1

